storage_service/locations/models/s3.py

```diff
@@ -21,68 +22,114 @@
+            error_code = e.response['Error']['Code']
+            if error_code == '404':
+                client.create_bucket(Bucket=self.bucket_name)
 
     @property
-    def client(self):
-        if self._client is None:
-            self._client = boto3.client(
-                service_name="s3",
-                endpoint_url=self.endpoint_url,
-                aws_access_key_id=self.access_key_id,
-                aws_secret_access_key=self.secret_access_key,
-                region_name=self.region,
-            )
-        return self._client
+    def bucket_name(self):
+        return self.s3_bucket or self.space_id
 
     @property
-    def resource(self):
-        if self._resource is None:
-            self._resource = boto3.resource(
-                service_name="s3",
-                endpoint_url=self.endpoint_url,
-                aws_access_key_id=self.access_key_id,
-                aws_secret_access_key=self.secret_access_key,
-                region_name=self.region,
-            )
+    def s3_resource(self):
+        if not hasattr(self, '_resource'):
+            if self.aws_access_key_id and self.aws_secret_access_key:
+
+                sts_client = boto3.client(
+                    service_name='sts',
+                    aws_access_key_id=self.aws_access_key_id,
+                    aws_secret_access_key=self.aws_secret_access_key,
+                )
+
+                assumed_role = sts_client.assume_role(
+                    RoleArn=self.aws_assumed_role,
+                    RoleSessionName='storage-session',
+                )
+                credentials = assumed_role['Credentials']
+
+                self._resource = boto3.resource(
+                    service_name='s3',
+                    endpoint_url=self.s3_endpoint_url,
+                    region_name=self.s3_region,
+                    aws_access_key_id=credentials['AccessKeyId'],
+                    aws_secret_access_key=credentials['SecretAccessKey'],
+                    aws_session_token=credentials['SessionToken'],
+                )
+            else:
+                self._resource = boto3.resource(
+                    service_name='s3',
+                    endpoint_url=self.s3_endpoint_url,
+                    region_name=self.s3_region,
+                )
+
         return self._resource
 
-    def _ensure_bucket_exists(self):
-        self.client.create_bucket(Bucket=self._bucket_name())
 
-    def _bucket_name(self):
-        return self.space_id
+
+class S3(S3SpaceModelMixin):
+    space = models.OneToOneField('Space', to_field='uuid')
+
+    class Meta(S3SpaceModelMixin.Meta):
+        verbose_name = _("S3")
+
+    ALLOWED_LOCATION_PURPOSE = [
+        Location.AIP_STORAGE,
+        Location.TRANSFER_SOURCE,
+    ]
 
     def browse(self, path):
         # strip leading slash on path
+        LOGGER.debug('Browsing %s on S3 storage' % path)
+        LOGGER.debug('Bucket: %s' % self.bucket_name)
```

Comment:
nit: prefer lazy logging, i.e.

```python
LOGGER.debug('Browsing %s on S3 storage', path)
LOGGER.debug('Bucket: %s', self.bucket_name)
```

Saves Python doing the string interpolation if it doesn't use the log.

Also, you could compress this into a single log line as an S3 URI:

```python
LOGGER.debug("Browsing s3://%s/%s", self.bucket_name, path)
```

But that’s a matter of preference and definitely not required!

Reply:
minor: also you’ve separated the comment from the code that uses it (L130–3). Arguably it’s not a very useful comment, but even less so when separated from the code it refers to!

Reply:
:+1: You're right, I don't think it was a very useful comment to begin with. I'll take it out.